game/scripts/enemies/TargetComputer.gd

```diff
@@ -1,32 +1,132 @@
+	var index_right: int = index + 1
+	var index_left: int = index - 1
+	if index_right >= n_candidates:
+		index_right = 0
+	if 	index_left < 0:
+		index_left = -1
+	return [index_left, index_right]
+	
 
 func compute_target(body_position: Vector2):
-#	print(target_detector.has_target())
-	var target_vector = null
-	var avoid_vector = null
-	var target_position = null
-	var avoid_position = null
-	if target_detector.has_target():
+	"""Computes the optimal path among n_candidate vector options via dot prod.
+	
+	This computes the dot product between all candidates and the target and
+	avoid locations. It then subtracts the avoid vector candidates from the 
+	attact vector candidates and weights the options by distance.
+	
+	The vector with the highest score is then selected and an offset position
+	is returned. An improvement would be to just have this return a direction
+	and have the parent Enemy figure out the specific location. 
+	"""
+	
+	# Each of these should likely be its own function call when switching to 
+	# potentially multiple targets / avoids.
+	has_target = target_detector.has_target()
+	if has_target:
 		target_position = target_detector.target.global_position
-		target_vector = body_position.direction_to(target_position)	
+		var target_vector = body_position.direction_to(target_position)	
+		target_scores = candidate_vectors.map(func vdot(cv): return cv.dot(target_vector))
+		target_weight = 1.0 / body_position.distance_to(target_position)
+	else: 
+		target_weight = 0
+		zero_target_weights()
+	
+	has_avoid = avoid_detector.has_target()
 	if avoid_detector.has_target():
 		avoid_position = avoid_detector.target.global_position
-		avoid_vector = body_position.direction_to(avoid_position)
-	if target_vector != null and avoid_vector != null:
-		print("DOT: ", target_vector.dot(avoid_vector))
-#		print("CROSS: ", avoid_vector.cross(target_vector))
-		print("NORMAL: ", target_vector.direction_to(avoid_position))
-		return target_position
+		var avoid_vector = body_position.direction_to(avoid_position)
+		avoid_scores = candidate_vectors.map(func vdot(cv): return cv.dot(avoid_vector))
+		# There is probably a better value in here like a sqrt or some
+		# other function to apply.
+		avoid_weight = 1.0 / body_position.distance_to(avoid_position)
+	else:
+		avoid_weight = 0
+		zero_avoid_weights()
+	
+	if not has_target and not has_avoid:
+		return body_position
+	# Kind of a hack so that the enemies don't just run away if you're behind
+	# an obstacle. 
+	target_weight = max(target_weight, avoid_weight)
+	var combined_scores: Array = range(n_candidates).map(func(i): return bravery * target_weight * target_scores[i] - ((1 - bravery) * avoid_weight * avoid_scores[i]))
+	var best_index: int = combined_scores.find(combined_scores.max())
+	# TODO (michaelwiest): Find neighboring candidates. And form a weighted average
+	# among them to interpolate between candidates. 
+	var neighbor_indices: Array = get_neighbor_indices(best_index)
```

Comment:
So does this assume then that the neighbors (index-1 and index+1) are also valid candidates so you may want to choose an average of candidates near the max score?